diff --git a/modules/pico_icmp6.c b/modules/pico_icmp6.c
index fdf6a982..55b897ef 100644
--- a/modules/pico_icmp6.c
+++ b/modules/pico_icmp6.c
@@ -58,7 +58,9 @@ uint16_t pico_icmp6_checksum(struct pico_frame *f)
 static void pico_icmp6_ping_recv_reply(struct pico_frame *f);
 #endif
 
-static int pico_icmp6_send_echoreply(struct pico_frame *echo)
+
+
+static int __attribute__((section(".ramfunc"))) pico_icmp6_send_echoreply(struct pico_frame *echo)
 {
     struct pico_frame *reply = NULL;
     struct pico_icmp6_hdr *ehdr = NULL, *rhdr = NULL;
@@ -91,6 +93,10 @@ static int pico_icmp6_send_echoreply(struct pico_frame *echo)
     return 0;
 }
 
+int wrapper_pico_icmp6_send_echoreply(struct pico_frame *echo) {
+    return pico_icmp6_send_echoreply(echo);
+}
+
 static int pico_icmp6_process_in(struct pico_protocol *self, struct pico_frame *f)
 {
     struct pico_icmp6_hdr *hdr = (struct pico_icmp6_hdr *)f->transport_hdr;
diff --git a/modules/pico_icmp6.h b/modules/pico_icmp6.h
index b4490752..c6a2dabf 100644
--- a/modules/pico_icmp6.h
+++ b/modules/pico_icmp6.h
@@ -323,4 +323,6 @@ int pico_icmp6_frag_expired(struct pico_frame *f);
 uint16_t pico_icmp6_checksum(struct pico_frame *f);
 int pico_icmp6_router_advertisement(struct pico_device *dev, struct pico_ip6 *dst);
 
+int wrapper_pico_icmp6_send_echoreply(struct pico_frame *echo);
+
 #endif
diff --git a/modules/pico_ipv6.c b/modules/pico_ipv6.c
index 922b679c..42756915 100644
--- a/modules/pico_ipv6.c
+++ b/modules/pico_ipv6.c
@@ -607,7 +607,7 @@ static int pico_ipv6_process_routing(struct pico_ipv6_exthdr *routing, struct pi
 
 #define IP6FRAG_MORE(x) ((x & 0x0001))
 
-static int pico_ipv6_process_destopt(struct pico_ipv6_exthdr *destopt, struct pico_frame *f, uint32_t opt_ptr)
+static int __attribute__((section(".ramfunc"))) pico_ipv6_process_destopt(struct pico_ipv6_exthdr *destopt, struct pico_frame *f, uint32_t opt_ptr)
 {
     uint8_t *option = NULL;
     uint8_t len = 0, optlen = 0;
@@ -656,6 +656,10 @@ static int pico_ipv6_process_destopt(struct pico_ipv6_exthdr *destopt, struct pi
     return 0;
 }
 
+int wrapper_pico_ipv6_process_destopt(struct pico_ipv6_exthdr *destopt, struct pico_frame *f, uint32_t opt_ptr) {
+    return pico_ipv6_process_destopt(destopt, f, opt_ptr);
+}
+
 static int pico_ipv6_check_headers_sequence(struct pico_frame *f)
 {
     struct pico_ipv6_hdr *hdr = (struct pico_ipv6_hdr *)f->net_hdr;
diff --git a/modules/pico_ipv6.h b/modules/pico_ipv6.h
index 56ddf6a7..35f0f0cc 100644
--- a/modules/pico_ipv6.h
+++ b/modules/pico_ipv6.h
@@ -179,4 +179,8 @@ int pico_ipv6_mcast_leave(struct pico_ip6 *mcast_link, struct pico_ip6 *mcast_gr
 struct pico_ipv6_link *pico_ipv6_get_default_mcastlink(void);
 
 int pico_ipv6_is_null_address(struct pico_ip6 *ip6);
+
+
+int wrapper_pico_ipv6_process_destopt(struct pico_ipv6_exthdr *destopt, struct pico_frame *f, uint32_t opt_ptr);
+
 #endif
